### PR TITLE
OPENEUROPA-1615: Implements a request serializer and update tests syntax.

### PR DESCRIPTION
**Pull request by drupol**
### Description

[Insert description here]

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: src/Serializer/RequestsSerializer.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+use Symfony\Component\Serializer\Encoder\YamlEncoder;
+use Symfony\Component\Serializer\Normalizer\ArrayDenormalizer;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * Class RequestsSerializer.
+ */
+final class RequestsSerializer implements SerializerInterface
```
Comment (ademarco) [inline src/Serializer/RequestsSerializer.php]: I can't think of any reason why this class should be final, I guess services can be theoretically extended.

File: tests/fixtures/requests/request-test.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        -
+          contacts:
+            contact:
+              -
+                UserId: '1'
+                RoleCode: AUTHOR
+              -
+                UserId: '2'
+                RoleCode: EDITOR
+          generalInfo:
```
Comment (ademarco) [inline tests/fixtures/requests/request-test.yml]: I'm a bit concerned about having different capitalization here, I mean:  `contacts` vs `UserId`, like this array will basically be the way external software will construct the request. Would it be too much of a mess to make this all consistent? Like I'd go for `Contacts`, for example, so camel case everywhere.

File: tests/Unit/PropertyInfo/Extractor/EpoetryReflectionExtractorTest.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+final class EpoetryReflectionExtractorTest extends TestCase
+{
+    /**
+     * Data provider.
+     *
+     * @return array
+     */
+    public function getTypesCases()
+    {
+        return $this->getFixture(__DIR__ . '/../../../fixtures/epoetry-reflection-extractor.yml');
```
Comment (ademarco) [inline tests/Unit/PropertyInfo/Extractor/EpoetryReflectionExtractorTest.php]: We can remove the `epoetry-` from the file name, I guess this is a leftover from when ePoetry was part of the class name, the extractor is called `MappingReflectionExtractor` so we should reflect this here.

Also I usually map the class namespace in the fixture directories, so this could be `property-info\Extractor\mapping-reflection-extractor.yml`

File: tests/Requests/MiddlewareTest.php [hunk 4]
```diff
@@ -5,29 +5,37 @@
...
+
     /**
      * Test Proxy Ticket in request.
+     *
+     * @dataProvider responseParsingCases
+     *
+     * @param mixed $input
      */
-    public function testResponseParsing()
+    public function testResponseParsing($input)
```
Comment (ademarco) [inline tests/Requests/MiddlewareTest.php]: Type hinting missing.

File: tests/Requests/MiddlewareTest.php [hunk 5]
```diff
@@ -5,29 +5,37 @@
...
     {
         // Generate request.
-        $generalInfo = new RequestGeneralInfoIn();
-        $generalInfo->setTitle('Test');
-        $linguisticRequestIn = new LinguisticRequestIn();
-        $linguisticRequestIn->setGeneralInfo($generalInfo);
-        $createRequests = new CreateRequests();
-        $createRequests->setLinguisticRequest($linguisticRequestIn);
+        $createRequests = RequestsSerializer::fromArray($input['request'], CreateRequests::class);
```
Comment (ademarco) [inline tests/Requests/MiddlewareTest.php]: We shouldn't use the same fixture in two different tests. We did not notice the problem with this because we are not bundling a fixture with both input and expected output.

I'll use the previous library as an example. This [component fixture](https://github.com/ec-europa/oe-poetry-client/blob/master/tests/src/fixtures/arrays/components/details.yml#L1) has both "input" and "output", meaning initial array with which to build a request and the assertions to be ran to test that everything works correctly. 

With should always have both in an entry so that we can run different "test scenarios" on the same tests, that's the ultimate scope of having data providers.

Comment (ademarco) [reply tests/Requests/MiddlewareTest.php]: Actually in this test we should have a raw XML as input and assert the object values, like we do [here](https://github.com/ec-europa/oe-poetry-client/blob/master/tests/src/fixtures/status.yml#L43) on the old library.

File: tests/Requests/RequestTest.php [hunk 6]
```diff
@@ -5,61 +5,39 @@
...
-        $linguisticRequestIn = new LinguisticRequestIn();
-        $linguisticRequestIn->setContacts($contacts)
-            ->setGeneralInfo($generalInfo);
-
-        $createRequests = new CreateRequests();
-        $createRequests->setLinguisticRequest($linguisticRequestIn);
+        $createRequests = RequestsSerializer::fromArray(
+            $input['request'],
+            CreateRequests::class
+        );
```
Comment (ademarco) [inline tests/Requests/RequestTest.php]: See comment above, we should have the expectations bundled together with the "input" test data, then we need to run assertions on those.

File: composer.json [hunk 7]
```diff
@@ -15,7 +15,10 @@
         "php-http/message": "^1.6",
         "php-http/message-factory": "^1.0",
         "phpro/soap-client": "^0.7.11",
-        "psr/http-message": "^1.0"
+        "psr/http-message": "^1.0",
+        "symfony/property-access": "^4.2",
+        "symfony/property-info": "^4.2",
+        "symfony/serializer": "^4.2"
```
Comment (ademarco) [inline composer.json]: This would not work with Drupal 8 since Drupal 8 is still on 3.x, we should have something like `"^3|^4"`.

File: src/Serializer/RequestsSerializer.php (hunk 1 above)
Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: I think that most of the classes we produce should be using the keyword 'final'.
This class `RequestSerializer` shouldn't be extended. The flexibility should come from good abstractions, not from inheritance. 

This library provides 1 way to create requests, **that** shouldn't be altered.
I doubt that one would change that anyway. However, if someone wants to alter it, then they can do it and use proper dependency injection mechanism.

This '`final`' word forces users to inject the service in their own class properly.

Example, instead of doing:

```php
<?php

namespace Foo\Bar;

class MyCustomRequestSerializer extends RequestSerializer {
  // Stuff.
}
```

They should do:

```php
<?php

namespace Foo\Bar;

class MyCustomRequestSerializer implements SerializerInterface {
  public function __construct(RequestsSerializer $serializer) {
    $this->serializer = $serializer;
  }
}
```

File: tests/fixtures/requests/request-test.yml (hunk 2 above)
Comment (drupol) [reply tests/fixtures/requests/request-test.yml]: Fixed by adding a Camel Case To Snake converter in the serializer.

File: tests/Unit/PropertyInfo/Extractor/EpoetryReflectionExtractorTest.php (hunk 3 above)
Comment (drupol) [reply tests/Unit/PropertyInfo/Extractor/EpoetryReflectionExtractorTest.php]: Fixed.

File: tests/Requests/MiddlewareTest.php (hunk 4 above)
Comment (drupol) [reply tests/Requests/MiddlewareTest.php]: Fixed.

File: tests/Requests/MiddlewareTest.php (hunk 5 above)
Comment (drupol) [reply tests/Requests/MiddlewareTest.php]: I was also a bit concerned about this, but @voidtek and @dxvargas told me that it's fine to use the same feature. Could you please check with them what to do ?

File: composer.json (hunk 7 above)
Comment (drupol) [reply composer.json]: Very good point.

Comment (drupol) [reply composer.json]: Fixed.

File: tests/Requests/MiddlewareTest.php (hunk 5 above)
Comment (ademarco) [reply tests/Requests/MiddlewareTest.php]: We were able to use the same feature because we where not pairing it with assertions, while we should have done that.

Comment (drupol) [reply tests/Requests/MiddlewareTest.php]: So, should we fix this in the scope of this ticket ?

File: src/Serializer/RequestsSerializer.php (hunk 1 above)
Comment (ademarco) [reply src/Serializer/RequestsSerializer.php]: Ok I see your point, we can keep it.

Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: Cool :-)

File: tests/Requests/MiddlewareTest.php (hunk 5 above)
Comment (drupol) [reply tests/Requests/MiddlewareTest.php]: Fixed.

File: tests/Requests/RequestTest.php (hunk 6 above)
Comment (drupol) [reply tests/Requests/RequestTest.php]: Fixed.

File: src/PropertyInfo/Extractor/MappingReflectionExtractor.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+use OpenEuropa\EPoetry\Type\ContactPersonIn;
+use OpenEuropa\EPoetry\Type\Contacts;
+use OpenEuropa\EPoetry\Type\CreateRequests;
+use OpenEuropa\EPoetry\Type\LinguisticRequestIn;
+use Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Class MappingReflectionExtractor.
```
Comment (ademarco) [inline src/PropertyInfo/Extractor/MappingReflectionExtractor.php]: We should add a description of the scope of this class, what it does, etc.

File: src/PropertyInfo/Extractor/MappingReflectionExtractor.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ */
+class MappingReflectionExtractor extends ReflectionExtractor
+{
+    /**
+     * The mapping list.
```
Comment (ademarco) [inline src/PropertyInfo/Extractor/MappingReflectionExtractor.php]: We should add instructions on how to add new entries in this property, ideally contributors should be able to understand how to extend this by simply reading the description.

File: src/Serializer/RequestsSerializer.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+use OpenEuropa\EPoetry\Type\CreateRequests;
+use Symfony\Component\Serializer\Encoder\YamlEncoder;
+use Symfony\Component\Serializer\NameConverter\CamelCaseToSnakeCaseNameConverter;
+use Symfony\Component\Serializer\Normalizer\ArrayDenormalizer;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+use Symfony\Component\Serializer\Serializer;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * Class RequestsSerializer.
```
Comment (ademarco) [inline src/Serializer/RequestsSerializer.php]: We should add a description of the scope of this class, what it does, etc.

File: src/Serializer/RequestsSerializer.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                new ArrayDenormalizer(),
+                new ObjectNormalizer(
+                    null,
+                    new CamelCaseToSnakeCaseNameConverter(),
+                    null,
+                    new MappingReflectionExtractor()
+                ),
+            ],
+            [
+                new YamlEncoder(),
```
Comment (ademarco) [inline src/Serializer/RequestsSerializer.php]: Do we need to set an encoder here? And shouldn't this class extend `Symfony\Component\Serializer\Serializer`?

File: src/Serializer/RequestsSerializer.php [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deserialize($data, $type, $format, array $context = [])
+    {
+        return $this->serializer->deserialize($data, $type, $format, $context);
+    }
```
Comment (ademarco) [inline src/Serializer/RequestsSerializer.php]: We could remove all these proxy methods if we extend `Symfony\Component\Serializer\Serializer`.

File: tests/AbstractTest.php [hunk 13]
```diff
@@ -22,12 +23,18 @@ abstract class AbstractTest extends TestCase
...
+     */
+    protected $expressionLanguage;
+
     /**
      * {@inheritdoc}
      */
     protected function setUp()
     {
         $this->httpClient = new Client();
+        $this->expressionLanguage = new ExpressionLanguage();
```
Comment (ademarco) [inline tests/AbstractTest.php]: We could [register here functions](https://symfony.com/doc/current/components/expression_language/extending.html#registering-functions) that we might use in the assertions (see comment below for more info).

We can easily register PHP function in the following way:

```php
        $this->expressionLanguage = new ExpressionLanguage();
        $this->expressionLanguage->addFunction(ExpressionFunction::fromPhp('count'));
```

File: tests/Middleware/MiddlewareTest.php [hunk 14]
```diff
@@ -2,36 +2,43 @@
...
+final class MiddlewareTest extends AbstractRequestTest
 {
+    /**
+     * Data provider.
+     *
+     * @return mixed
+     */
+    public function responseParsingCases()
+    {
+        return $this->getFixture('request-test.yml');
```
Comment (ademarco) [inline tests/Middleware/MiddlewareTest.php]: We should rename this fixture to match the test it provides data for, I'd suggest something like: `response-parsing-test.yml`.

File: tests/Requests/RequestTest.php [hunk 15]
```diff
@@ -5,220 +5,115 @@
...
-        $contact = new ContactPersonIn();
-        $contact->setUserId('1');
-        $contact->setRoleCode('AUTHOR');
-        $contacts->addContact($contact);
-
-        $contact = new ContactPersonIn();
-        $contact->setUserId('2');
-        $contact->setRoleCode('EDITOR');
-        $contacts->addContact($contact);
+        return $this->getFixture('request-test.yml');
```
Comment (ademarco) [inline tests/Requests/RequestTest.php]: Are we using the same fixture for two different tests? This might work for now but it will quickly make tests not maintainable, we should separate the two.

File: tests/Requests/RequestTest.php [hunk 16]
```diff
@@ -5,220 +5,115 @@
...
+            'response' => $response,
+        ];
+
+        foreach ($expectations['assertions'] as $type => $expectationsTypes) {
+            foreach ($expectationsTypes as $expression => $expectation) {
+                $this->{$type}(
+                    $expectation,
+                    $this->expressionLanguage->evaluate(
+                        $expression,
+                        $values
```
Comment (ademarco) [inline tests/Requests/RequestTest.php]: We can actually evaluate expressions without the need of PHPUnit methods.

File: tests/fixtures/requests/response-parsing.yml [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+      <SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://eu.europa.ec.dgt.epoetry">
+      <SOAP-ENV:Body>
+      <ns1:createRequests>
+      <linguisticRequest><generalInfo><title>Test</title><internalReference>1</internalReference><internalTechnicalId>1</internalTechnicalId><requestedDeadline>2020-02-02T00:00:00+00:00</requestedDeadline><sensitive>false</sensitive><documentToBeAdopted>true</documentToBeAdopted><decideReference>decideReference</decideReference><sentViaRUE>true</sentViaRUE><destinationCode>PUBLIC</destinationCode><slaAnnex></slaAnnex><slaCommitment></slaCommitment><comment></comment><onBehalfOf></onBehalfOf><accessibleTo></accessibleTo></generalInfo>
+      <contacts><contact userId="1" roleCode="AUTHOR"/><contact userId="2" roleCode="EDITOR"/></contacts>
+      </linguisticRequest>
+      </ns1:createRequests>
+      </SOAP-ENV:Body>
+      </SOAP-ENV:Envelope>
+    assertions:
```
Comment (ademarco) [inline tests/fixtures/requests/response-parsing.yml]: We can convert these assertions in a simple list of expressions, such as:

```yml
    assertions:
      - 'count(response.getReturn()) === 2'
      - 'response.getReturn()[0].getStatusCode() === ENV'
```
This will make tests easier to write and more extensible. Note that `count()` should be registered as indicated in my previous comment.

File: composer.json [hunk 18]
```diff
@@ -15,13 +15,17 @@
         "php-http/message": "^1.6",
         "php-http/message-factory": "^1.0",
         "phpro/soap-client": "^0.7.11",
-        "psr/http-message": "^1.0"
+        "psr/http-message": "^1.0",
+        "symfony/property-access": "^3|^4",
+        "symfony/property-info": "^3.2|^4",
+        "symfony/serializer": "^3.2|^4"
```
Comment (dxvargas) [inline composer.json]: We should use 2 parallel pipe lines `||`

Comment (drupol) [reply composer.json]: omg... how the... what ?!

Comment (drupol) [reply composer.json]: One valid and good reason to user `composer-normalizer` from @localheinz :-)

File: src/Serializer/RequestsSerializer.php (hunk 11 above)
Comment (voidtek) [reply src/Serializer/RequestsSerializer.php]: Can we add the encoder XML?

File: src/Serializer/RequestsSerializer.php (hunk 12 above)
Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: We already discussed this earlier. In this case we would prefer composition over inheritance, this is why the class is final.

File: tests/Middleware/MiddlewareTest.php (hunk 14 above)
Comment (drupol) [reply tests/Middleware/MiddlewareTest.php]: :+1:

File: src/Serializer/RequestsSerializer.php (hunk 10 above)
Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: Fixed.

File: src/Serializer/RequestsSerializer.php (hunk 11 above)
Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: Fixed.

File: tests/Requests/RequestTest.php (hunk 15 above)
Comment (drupol) [reply tests/Requests/RequestTest.php]: Fixed.

File: composer.json (hunk 18 above)
Comment (drupol) [reply composer.json]: Fixed.

File: src/PropertyInfo/Extractor/MappingReflectionExtractor.php (hunk 8 above)
Comment (drupol) [reply src/PropertyInfo/Extractor/MappingReflectionExtractor.php]: Done.

File: src/PropertyInfo/Extractor/MappingReflectionExtractor.php (hunk 9 above)
Comment (drupol) [reply src/PropertyInfo/Extractor/MappingReflectionExtractor.php]: Fixed.

File: tests/AbstractTest.php (hunk 13 above)
Comment (drupol) [reply tests/AbstractTest.php]: Fixed.

File: tests/Requests/RequestTest.php (hunk 16 above)
Comment (drupol) [reply tests/Requests/RequestTest.php]: Fixed.

File: tests/fixtures/requests/response-parsing.yml (hunk 17 above)
Comment (drupol) [reply tests/fixtures/requests/response-parsing.yml]: Fixed.

File: tests/Middleware/MiddlewareTest.php [hunk 19]
```diff
@@ -15,50 +15,61 @@
...
+        }
+
         $client->createRequests(new CreateRequests());
 
-        // Perform request.
-        $request = $client->debugLastSoapRequest()['request'];
+        $values['request'] = $client->debugLastSoapRequest()['request'];
 
-        $this->assertContains('ecas:ProxyTicket: DESKTOP_PT-21-9fp9', $request['headers'], 'Request XML header malformed, missing proxy ticket.');
+        foreach ($expectations['assertions'] as $expression) {
```
Comment (ademarco) [inline tests/Middleware/MiddlewareTest.php]: This should be moved into its own method as it's repeated 3 times in different test classes.

Comment (drupol) [reply tests/Middleware/MiddlewareTest.php]: Fixed.

File: tests/fixtures/Middleware/proxy-ticket-test.yml [hunk 20]
```diff
@@ -0,0 +1,229 @@
+-
+  ticket: DESKTOP_PT-21-9fp9
+  xml: |
```
Comment (dxvargas) [inline tests/fixtures/Middleware/proxy-ticket-test.yml]: This xml is repeated. Here, in `tests/fixtures/Requests/request-send-test.yml` and in `tests/fixtures/Requests/response-parse-test.yml`. And may it's going to be needed more times?

It should be a fixture _per se_ and reused when necessary.

On the other hand, it's not really relevant to the test, is it? If it's a temporary work around then it's not a big problem. But then we need a ticket to work on this, it can't stay as is.

Comment (dxvargas) [inline tests/fixtures/Middleware/proxy-ticket-test.yml]: It's weird to mix XML inside yml.
Maybe we could have a path to a xml file:
xml: path/to/fixtures/xml/gimme.xml

File: tests/fixtures/Requests/request-send-test.yml [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+      <SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns1="http://eu.europa.ec.dgt.epoetry">
+      <SOAP-ENV:Body>
+      <ns1:createRequests>
+      <linguisticRequest><generalInfo><title>Test</title><internalReference>1</internalReference><internalTechnicalId>1</internalTechnicalId><requestedDeadline>2020-02-02T00:00:00+00:00</requestedDeadline><sensitive>false</sensitive><documentToBeAdopted>true</documentToBeAdopted><decideReference>decideReference</decideReference><sentViaRUE>true</sentViaRUE><destinationCode>PUBLIC</destinationCode><slaAnnex></slaAnnex><slaCommitment></slaCommitment><comment></comment><onBehalfOf></onBehalfOf><accessibleTo></accessibleTo></generalInfo>
+      <contacts><contact userId="1" roleCode="AUTHOR"/><contact userId="2" roleCode="EDITOR"/></contacts>
+      </linguisticRequest>
+      </ns1:createRequests>
+      </SOAP-ENV:Body>
+      </SOAP-ENV:Envelope>
+    assertions:
```
Comment (dxvargas) [inline tests/fixtures/Requests/request-send-test.yml]: For me it's weird to see assertions inside the fixtures. I think these should live in the tests file. Fixtures are something that we provide to mock data and to compare against. So for me we shouldn't have the tests here.
I also have doubts about the expectations.

File: tests/fixtures/Middleware/proxy-ticket-test.yml (hunk 20 above)
Comment (drupol) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: I think we discussed with @ademarco already about this.
We would like to have in the fixture files all the necessary input data and the expectations.
Obviously, this might lead to some duplicated data.

Comment (drupol) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: This is also possible... but then it will be more fixtures files to handle.
I'll leave you guys decide and I'll update accordingly.

File: tests/fixtures/Requests/request-send-test.yml (hunk 21 above)
Comment (drupol) [reply tests/fixtures/Requests/request-send-test.yml]: This is the whole point of this PR... if using expression language (_and thus, assertions in the yaml_), then we had to have this discussion before doing the PR  :(

File: composer.json [hunk 22]
```diff
@@ -13,26 +13,26 @@
...
         "phpro/soap-client": "^0.7.11",
-        "psr/http-message": "^1.0"
+        "psr/http-message": "^1.0",
+        "symfony/property-access": "^3 || ^4",
+        "symfony/property-info": "^3.2 || ^4",
+        "symfony/serializer": "^3.2 || ^4"
     },
     "require-dev": {
-        "cweagans/composer-patches": "~1.6",
+        "cweagans/composer-patches": "^1.6",
```
[the remark this answers is not in the document]
Comment (drupol) [reply composer.json]: You're perfectly right. Fixed.

File: src/Serializer/RequestsSerializer.php [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+     * Convert a PHP array into an object.
+     *
+     * @param array $input
+     *   The PHP array
+     * @param string $type
+     *   The object type
+     *
+     * @throws \Symfony\Component\Serializer\Exception\ExceptionInterface
+     *
+     * @return object
```
[the remark this answers is not in the document]
Comment (drupol) [reply src/Serializer/RequestsSerializer.php]: Fixed.

File: tests/Requests/RequestTest.php [hunk 24]
```diff
@@ -5,220 +5,92 @@
...
+        return $this->getFixture('response-parse-test.yml');
+    }
 
-        $createRequests = new CreateRequests();
-        $createRequests->setLinguisticRequest([$linguisticRequestIn]);
+    /**
+     * Test a SOAP request.
+     *
+     * @dataProvider requestSendingCases
+     *
```
[the remark this answers is not in the document]
Comment (drupol) [reply tests/Requests/RequestTest.php]: How do you see that ?
It wasn't there before!
It's not that I don't trust you, but I would like to be able to spot issues like that as well when doing review! I'm curious :)

File: tests/fixtures/Requests/response-parse-test.yml [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+            DecideReference: decideReference
+            SentViaRUE: true
+            DestinationCode: PUBLIC
+            SlaAnnex: ''
+            SlaCommitment: ''
+            Comment: ''
+            OnBehalfOf: ''
+            AccessibleTo: ''
+  expectations:
+    request: |
```
Comment (dxvargas) [inline tests/fixtures/Requests/response-parse-test.yml]: This is not being used anywhere...

File: tests/fixtures/Requests/request-send-test.yml (hunk 21 above)
Comment (ademarco) [reply tests/fixtures/Requests/request-send-test.yml]: Assertions and fixtures goes together, you assert against a fixture, this is used everywhere, in data providers you pass the data of the test and the expected result so that you can use the same test to test more than one scenario.

Look at [data provider doc](https://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers) to have an example of it:

```php
<?php
class DataTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider provider
     */
    public function testAdd($a, $b, $c)
    {
        $this->assertEquals($c, $a + $b);
    }

    public function provider()
    {
        return array(
          array(0, 0, 0),
          array(0, 1, 1),
          array(1, 0, 1),
          array(1, 1, 3)
        );
    }
}
?>
```

File: tests/fixtures/Middleware/proxy-ticket-test.yml (hunk 20 above)
Comment (ademarco) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: This will make fixtures hard to read, while if we keep all data in the same place it's easy to figure out what every test scenario does.

Comment (ademarco) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: Yes duplicated data is fine but if on the same test we need always the same input then that's not supposed to be part of the fixture and can be moved within the test itself, thus considering it as test setup.

File: tests/Requests/RequestTest.php [hunk 26]
```diff
@@ -5,220 +5,92 @@
...
+            $expectations['assertions'],
+            $values
+        );
     }
 
     /**
      * Test parsing a SOAP response.
+     *
+     * @dataProvider responseParsingCases
+     *
```
[the remark this answers is not in the document]
Comment (ademarco) [reply tests/Requests/RequestTest.php]: In a test there is no point of documenting what exceptions are going to be thrown.

File: tests/Requests/RequestTest.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (ademarco) [reply tests/Requests/RequestTest.php]: In a test there is no point of documenting what exceptions are going to be thrown.

File: tests/fixtures/Requests/response-parse-test.yml (hunk 25 above)
Comment (ademarco) [reply tests/fixtures/Requests/response-parse-test.yml]: We should removed it then, it might be a leftover.

Comment (drupol) [reply tests/fixtures/Requests/response-parse-test.yml]: Fixed.

Comment (drupol) [reply tests/fixtures/Requests/response-parse-test.yml]: Fixed.

File: tests/Requests/RequestTest.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (drupol) [reply tests/Requests/RequestTest.php]: @ademarco PHPStorm adds them automatically. On my side it wasn't added because for some reason I had disabled the feature.
That said, there or not, I don't think it's problem.
Don't you ?

[the remark this answers is not in the document]
Comment (dxvargas) [reply tests/Requests/RequestTest.php]: Yes, it's ugly to see, PHPStorm underlines the all doc block because this missing missing.

File: tests/fixtures/Middleware/proxy-ticket-test.yml (hunk 20 above)
Comment (dxvargas) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: Ok then, let's keep it as is.

Comment (dxvargas) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: Here and in most of the cases xml for responses can be simplified to just ensure we get a createRequestsResponse in response. Like this:

```
     <?xml version='1.0' encoding='UTF-8'?>
      <S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/">
          <S:Body>
              <ns0:createRequestsResponse xmlns:ns0="http://eu.europa.ec.dgt.epoetry">
              </ns0:createRequestsResponse>
          </S:Body>
      </S:Envelope>
```

But anyway, my point was that it shouldn't be repeated but if you consider that duplicated is ok I rest my case.

File: tests/fixtures/Requests/request-send-test.yml (hunk 21 above)
Comment (dxvargas) [reply tests/fixtures/Requests/request-send-test.yml]: @ademarco I never saw the tests being added in fixtures (while fixtures being added in tests like the example you're showing is common.
But ok thanks for clarifying.

File: tests/fixtures/Middleware/proxy-ticket-test.yml [hunk 27]
```diff
@@ -1,27 +1,13 @@
...
   ticket: ''
-  xml: |
-    <?xml version='1.0' encoding='UTF-8'?>
-    <S:Envelope xmlns:S="http://schemas.xmlsoap.org/soap/envelope/">
-        <S:Body>
-            <ns0:createRequestsResponse xmlns:ns0="http://eu.europa.ec.dgt.epoetry">
-            </ns0:createRequestsResponse>
-        </S:Body>
-    </S:Envelope>
+  xml: request-response.xml
```
Comment (ademarco) [inline tests/fixtures/Middleware/proxy-ticket-test.yml]: This doesn't need to be in the fixture, it can be part of the test setup as it's basically an empty response. Let's just dump the XML in the `MiddlewareTest` class somewhere so we get rid of unnecessary files.

Comment (drupol) [reply tests/fixtures/Middleware/proxy-ticket-test.yml]: Fixed.

**Comment by drupol** [comment]
Bummer !
